### PR TITLE
Skip documentation files without ``` when running markdown tests.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This should reduce the 'running 0 tests' noise in builds, and I believe this is a good heuristic for us to use.

cc @rust-lang/docs -- do we use the indented format for code blocks anywhere? Will we? If so, we shouldn't do this.

r? @alexcrichton

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Why not? :)

**Comment by kennytm** [comment]
```sh
$ rg -C5 --glob '*.md' '//[/!]    '
src/doc/book/first-edition/src/documentation.md
434-not actually pass as a test.
435-
436-```rust
437-/// ```rust,no_run
438-/// loop {
439:///     println!("Hello, world");
440-/// }
441-/// ```
442-# fn foo() {}
443-```
444-
```

It should be safe to go ahead. But there is no harm to check also ``///     `` and `//!     ` as well.

**Comment by Mark-Simulacrum** [comment]
I'm not sure I follow -- those presumably wouldn't be present in markdown files? Are they indications of code blocks? It doesn't feel to me like they are, but I could be wrong.

**Comment by alexcrichton** [comment]
Looks reasonable to me! My only worry is, as you commented above, not catching indented blocks or those separated by `~~~` (that's historical, right?).

That being said if our style is to only use backticks (which I think it is) then this looks good to me.

**Comment by Mark-Simulacrum** [comment]
Yeah, I want @steveklabnik to comment to make sure that we're good with just backticks, but after that I think we'll be good to go. I initially wanted to put this sort of functionality into rustdoc directly (calling it with multiple files which it'd process separately but print together in one list) but that seemed more complicated and I decided to wait on that at least until the next rustdoc.

File: src/bootstrap/check.rs [hunk 1]
```diff
@@ -376,6 +368,13 @@ pub fn error_index(build: &Build, compiler: &Compiler) {
 }
 
 fn markdown_test(build: &Build, compiler: &Compiler, markdown: &Path) {
+    let mut file = t!(File::open(markdown));
+    let mut contents = String::new();
+    t!(file.read_to_string(&mut contents));
+    if !contents.contains("```") {
```
Comment (steveklabnik) [inline src/bootstrap/check.rs]: this isn't the only way there are code blocks though; for example, four spaces is also considered a code block

**Review by steveklabnik — COMMENTED** [review]
I am not opposed to this idea, but it also doesn't fully cover all the cases yet.

**Comment by steveklabnik** [comment]
> That being said if our style is to only use backticks (which I think it is) then this looks good to me.

Yeah, it is; if everyone else is okay with just checking backticks, then I won't let the perfect be the enemy of the good 😄

**Comment by Mark-Simulacrum** [comment]
I could add something like \n followed by 4 spaces, but I feel like that heuristic wouldn't work too well due to indentation. I'm open to adding it if we want to though.

**Comment by alexcrichton** [comment]
@steveklabnik the failure mode here will be that if you write a doc example or a doc block *without* three ticks in a markdown file then it won't get tested (in just rust-lang/rust during `./x.py test`). Are you ok with that or do you think we should catch more cases?

@Mark-Simulacrum this is primarily done to cut down on all the output generated, right? Do you know how many files end up having 0 tests?

**Comment by Mark-Simulacrum** [comment]
I believe 417 files would be skipped with this check.

```
rg --files-without-match '```' | wc -l
```

**Comment by steveklabnik** [comment]
> @steveklabnik the failure mode here will be that if you write a doc example or a doc block without three ticks in a markdown file then it won't get tested (in just rust-lang/rust during ./x.py test). Are you ok with that or do you think we should catch more cases?

I think virtually everything uses the triple ticks, so it should be fine. Gonna bookmark this comment for someday when this randomly bites us, though 😆

**Comment by alexcrichton** [comment]
Heh ok! Let's get that sweet sweet thousands-of-less-lines-of-output from travis then!

@bors: r+

**Pull request merged**